tools/walletextension/ratelimiter/rate_limiter.go

```diff
@@ -0,0 +1,215 @@
+		for _, interval := range user.CurrentRequests {
+			if interval.End != nil && interval.End.After(cutoff) {
+				totalComputeTime += interval.End.Sub(interval.Start)
+			}
+		}
+	}
+	return uint32(totalComputeTime / time.Millisecond)
+}
+
+type RateLimiter struct {
+	mu                    sync.Mutex
+	users                 map[common.Address]*RateLimitUser
+	userComputeTime       uint32
+	window                uint32
+	maxConcurrentRequests uint32
+	totalRequests         uint64
+	rateLimitedRequests   uint64
+}
+
+// IncrementTotalRequests increments the total requests counter by 1 with thread safety.
+func (rl *RateLimiter) IncrementTotalRequests() {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	rl.totalRequests++
+}
+
+// IncrementRateLimitedRequests increments the total requests counter by 1 with thread safety.
+func (rl *RateLimiter) IncrementRateLimitedRequests() {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	rl.rateLimitedRequests++
+}
+
+// GetMaxConcurrentRequest returns the maximum number of concurrent requests allowed.
+func (rl *RateLimiter) GetMaxConcurrentRequest() uint32 {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	return rl.maxConcurrentRequests
+}
+
+// GetUserComputeTime returns the user compute time in milliseconds.
+func (rl *RateLimiter) GetUserComputeTime() uint32 {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	return rl.userComputeTime
+}
+
+func NewRateLimiter(rateLimitUserComputeTime uint32, rateLimitWindow uint32, concurrentRequestsLimit uint32) *RateLimiter {
+	rl := &RateLimiter{
+		users:                 make(map[common.Address]*RateLimitUser),
+		userComputeTime:       rateLimitUserComputeTime,
+		window:                rateLimitWindow,
+		maxConcurrentRequests: concurrentRequestsLimit,
+	}
+	go rl.logRateLimitedStats()
+	go rl.periodicPrune()
+	return rl
+}
+
+// Allow checks if the user is allowed to make a request based on the rate limit threshold
+// before comparing to the threshold also decays the score of the user based on the decay rate
+func (rl *RateLimiter) Allow(userID common.Address) (bool, uuid.UUID) {
+	// If the userComputeTime is 0, allow all requests (rate limiting is disabled)
+	if rl.GetUserComputeTime() == 0 {
+		return true, zeroUUID
+	}
+	// Increment the total requests counter for statistics
+	rl.IncrementTotalRequests()
+
+	// Check if the user has reached the maximum number of concurrent requests
+	if uint32(rl.CountOpenRequests(userID)) >= rl.GetMaxConcurrentRequest() {
+		rl.IncrementRateLimitedRequests()
+		return false, zeroUUID
+	}
+
+	// Check if user is in limits of rate limiting
+	userComputeTimeForUser := rl.SumComputeTime(userID)
+	if userComputeTimeForUser <= rl.userComputeTime {
+		requestUUID := rl.AddRequest(userID, RequestInterval{Start: time.Now()})
+		return true, requestUUID
+	}
+	rl.IncrementRateLimitedRequests()
+	return false, zeroUUID
+}
+
+// PruneRequests deletes all requests that have ended before the rate limiter's window.
+func (rl *RateLimiter) PruneRequests() {
```

Comment:
Be nice to know if this starts being slow, maybe a log line at the end if it's over some amount like:

```
timeTaken := time.Since(startTime)
if timeTaken > 1 * time.Second {
   log.Printf("PruneRequests completed in %s", timeTaken)
}
```

Bit worried the contention on the mutex lock could become troublesome. If it does I guess we can add a per-user rw-lock in the User object so pruning doesn't stop new requests on existing users being allowed.

Reply:
added logs if time taken > 1s. 

I am aware of mutex lock problems (created getters, increments and not locking it for too long).
Do you think we should try with this and later (if it becomes a problem) switch to per-user locking?

Reply:
Yeah I think it's fine for now all your locks looked good to me and might not be an issue at all. But something to consider if we start seeing unexplained slowness at some point.